### PR TITLE
feat: Arrow migration

### PR DESCRIPTION
**Pull request by yevgenypats**
Blocked by https://github.com/cloudquery/plugin-sdk/pull/724

but ready for initial review and discussion. I can do a short walkthrough for anyone up for review.

Apache arrow fork is here (We use it until [this](https://github.com/apache/arrow/pull/34454) is merged): https://github.com/cloudquery/arrow/tree/feat_extension_builder.

Some more notes:
- Right now we are just migrating the json writer/reader to use apache arrow so we can roll format by format and see if there are any real world issues before we roll this to everywhere instead of our own type system.

### CODE REVIEW

File: internal/cqarrow/arrow.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+func CQSchemaToArrow(table *schema.Table) *arrow.Schema {
+	fields := make([]arrow.Field, 0, len(table.Columns))
+	for _, col := range table.Columns {
+		fields = append(fields, CQColumnToArrowField(&col))
+	}
+	return arrow.NewSchema(fields, nil)
+}
+
+func CqTypesToRecord(mem memory.Allocator, c []schema.CQTypes, arrowSchema *arrow.Schema) arrow.Record {
```
Comment (disq) [inline internal/cqarrow/arrow.go:77]:
```suggestion
func CQTypesToRecord(mem memory.Allocator, c []schema.CQTypes, arrowSchema *arrow.Schema) arrow.Record {
```

File: internal/cqarrow/arrow.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	case schema.TypeString:
+		typ = arrow.BinaryTypes.String
+	case schema.TypeByteArray:
+		typ = arrow.BinaryTypes.Binary
+	case schema.TypeStringArray:
+		typ = arrow.ListOf(arrow.BinaryTypes.String)
+	case schema.TypeIntArray:
+		typ = arrow.ListOf(arrow.PrimitiveTypes.Int64)
+	case schema.TypeTimestamp:
+		typ = arrow.FixedWidthTypes.Timestamp_us
```
Comment (erezrokah) [inline internal/cqarrow/arrow.go:38]: Maybe add a comment that `us` means micros seconds (and not US timezone)

**Review by erezrokah — APPROVED** [review]
Looks good (didn't test). So the TDLR for this PR is that we use Apache arrow type system instead of ours (extending types they don't support)?

A walkthrough will be great, and also seeing them accept some of our extended types as contributions

File: internal/cqarrow/arrow.go (hunk 2 above)
Comment (disq) [reply internal/cqarrow/arrow.go:38]: arrow type is named badly, should've been `arrow.FixedWidthTypes.TimestampMicros`

File: internal/cqarrow/inet.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	if delim, ok := t.(json.Delim); !ok || delim != '[' {
+		return fmt.Errorf("fixed size binary builder must unpack from json array, found %s", delim)
+	}
+
+	return b.Unmarshal(dec)
+}
+
+// UUIDArray is a simple array which is a FixedSizeBinary(16)
```
Comment (hermanschaaf) [inline internal/cqarrow/inet.go:104]: It looks like all the comments in this file that refer to `FixedSizeBinary` or `UUID` need to be updated (they were copied from UUID)

File: internal/cqarrow/mac.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	if delim, ok := t.(json.Delim); !ok || delim != '[' {
+		return fmt.Errorf("fixed size binary builder must unpack from json array, found %s", delim)
+	}
+
+	return b.Unmarshal(dec)
+}
+
+// UUIDArray is a simple array which is a FixedSizeBinary(16)
```
Comment (hermanschaaf) [inline internal/cqarrow/mac.go:100]: Same here: It looks like all the comments in this file that refer to FixedSizeBinary or UUID need to be updated (they were copied from UUID)

File: internal/cqarrow/json.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		err := json.Unmarshal(arr.Value(i), &data)
+		if err != nil {
+			panic(fmt.Errorf("invalid json: %w", err))
+		}
+		return data
+	}
+	return nil
+}
+
+// JSONType is a simple extension type that represents a FixedSizeBinary(16)
```
Comment (hermanschaaf) [inline internal/cqarrow/json.go:180]: Probably not really FixedSizeBinary(16) for JSON :)

File: internal/cqarrow/json.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	if data != "json-serialized" {
+		return nil, fmt.Errorf("type identifier did not match: '%s'", data)
+	}
+	if !arrow.TypeEqual(storageType, &arrow.BinaryType{}) {
+		return nil, fmt.Errorf("invalid storage type for JSONType: %s", storageType.Name())
+	}
+	return NewJSONType(), nil
+}
+
+// UuidTypes are equal if both are named "uuid"
```
Comment (hermanschaaf) [inline internal/cqarrow/json.go:220]:
```suggestion
// Json types are equal if both have the same name
```

File: internal/cqarrow/inet_test.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+		mustParseInet("192.168.0.0/26"),
+		mustParseInet("192.168.0.0/27"),
+	}
+	b.AppendValues(values, []bool{true, true})
+
+	require.Equal(t, 6, b.Len(), "unexpected Len()")
+
+	a := b.NewArray()
+
+	// check state of builder after NewUUIDBuilder
```
Comment (hermanschaaf) [inline internal/cqarrow/inet_test.go:43]:
```suggestion
	// check state of builder after NewInetBuilder
```

File: internal/cqarrow/json_test.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+	require.Equal(t, 6, b.Len(), "unexpected Len()")
+
+	a := b.NewArray()
+
+	// check state of builder after NewUUIDBuilder
+	require.Zero(t, b.Len(), "unexpected ArrayBuilder.Len(), NewUUIDBuilder did not reset state")
+	require.Zero(t, b.Cap(), "unexpected ArrayBuilder.Cap(), NewUUIDBuilder did not reset state")
+	require.Zero(t, b.NullN(), "unexpected ArrayBuilder.NullN(), NewUUIDBuilder did not reset state")
+	require.Equal(t, `["{"a":1,"b":2}" (null) "{"c":3,"d":4}" (null) "{"e":5,"f":6}" "{"g":7,"h":8}"]`, a.String())
```
Comment (hermanschaaf) [inline internal/cqarrow/json_test.go:37]:
```suggestion
	// check state of builder after NewJSONBuilder
	require.Zero(t, b.Len(), "unexpected ArrayBuilder.Len(), NewJSONBuilder did not reset state")
	require.Zero(t, b.Cap(), "unexpected ArrayBuilder.Cap(), NewJSONBuilder did not reset state")
	require.Zero(t, b.NullN(), "unexpected ArrayBuilder.NullN(), NewJSONBuilder did not reset state")
	require.Equal(t, `["{"a":1,"b":2}" (null) "{"c":3,"d":4}" (null) "{"e":5,"f":6}" "{"g":7,"h":8}"]`, a.String())
```

File: internal/cqarrow/mac.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+func (MacType) Serialize() string { return "mac-serialized" }
+
+// Deserialize expects storageType to be FixedSizeBinaryType{ByteWidth: 16} and the data to be
+// "mac-serialized" in order to correctly create a MacType for testing deserialize.
+func (MacType) Deserialize(storageType arrow.DataType, data string) (arrow.ExtensionType, error) {
+	if data != "mac-serialized" {
+		return nil, fmt.Errorf("type identifier did not match: '%s'", data)
+	}
+	if !arrow.TypeEqual(storageType, &arrow.BinaryType{}) {
+		return nil, fmt.Errorf("invalid storage type for InetType: %s", storageType.Name())
```
Comment (hermanschaaf) [inline internal/cqarrow/mac.go:177]:
```suggestion
		return nil, fmt.Errorf("invalid storage type for MacType: %s", storageType.Name())
```

File: internal/cqarrow/arrow_test.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+	"testing"
+
+	"github.com/apache/arrow/go/v12/arrow"
+	"github.com/apache/arrow/go/v12/arrow/memory"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/cloudquery/plugin-sdk/testdata"
+	"github.com/google/go-cmp/cmp"
+)
+
+func DiffSchema(sc, o *arrow.Schema) (string, bool) {
```
Comment (hermanschaaf) [inline internal/cqarrow/arrow_test.go:15]: I think this is some dead code that can be removed?

File: internal/cqarrow/arrow_test.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+		{Name: "macaddr", Type: NewMacType(), Nullable: true},
+		{Name: "macaddr_array", Type: arrow.ListOf(NewMacType()), Nullable: true},
+	}, nil)
+
+	testTable := testdata.TestSourceTable("test_table")
+	arrowSchema := CQSchemaToArrow(testTable)
+	if diff := cmp.Diff(arrowSchema.String(), expecetdSchema.String()); diff != "" {
+		t.Errorf(diff)
+	}
+	// if diff, _ := DiffSchema(arrowSchema, expecetdSchema); diff != "" {
```
Comment (hermanschaaf) [inline internal/cqarrow/arrow_test.go:70]: (The only use of DiffSchema is here and it's commented out)

File: internal/cqarrow/json.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+	}
+	return b
+}
+
+func (b *JSONBuilder) Append(v any) {
+	if v == nil {
+		b.AppendNull()
+		return
+	}
+	bytes, err := json.Marshal(v)
```
Comment (hermanschaaf) [inline internal/cqarrow/json.go:33]: I'm wondering if we need to use json.Encoder here as well, like we had to in plugin-sdk (https://github.com/cloudquery/plugin-sdk/pull/714/files)?

**Review by hermanschaaf — APPROVED** [review]
LGTM - I tested it against the old json filetypes plugin for a couple of sources and the output was the byte-for-byte the same 👍 We may want to consider using `json.Encoder` but I think it's not a blocker

**Pull request merged**